### PR TITLE
Further work on lkt semantics

### PR DESCRIPTION
**Pull request by raph-amiard**

### CODE REVIEW

File: contrib/lkt/language/parser.py [hunk 1]
```diff
@@ -1923,28 +1949,37 @@ class GenericInstantiation(Expr):
...
+        )
+    )
 
     @langkit_property(public=True)
-    def designated_type():
+    def instantiated_decl():
         """
         Get the type designated by this instantiation.
 
         NOTE: for the moment we only have generic types, so that's enough. If
```
Comment (Roldak) [inline contrib/lkt/language/parser.py]: Remove the note?

File: contrib/lkt/language/parser.py [hunk 2]
```diff
@@ -1862,16 +1901,187 @@ def check_correctness():
...
+                .concat(S("` with `")).concat(expected_type.full_name)
+                .concat(S("`"))
+            )),
+
+            # Inference returned an error with message -> propagate it
+            Not(unsorted_actuals.error.is_null),
+            unsorted_actuals,
+
+            # We don't have all the actuals needed to infer -> return an error
+            unsorted_actuals.result.length
```
Comment (Roldak) [inline contrib/lkt/language/parser.py]: It seems to me like this check is too soft. Consider for example:
```
def mkPair[T](x: T) -> Pair[T, T]

def foo(): Pair[Int, Bool] = mkPair(1)
```
Am I right that `infer_actuals_impl` would return `[T -> Int, T -> Bool]` here? So, this check would pass and further down in `check_infer_actuals` we will take the first actual for `T`, so `Int`, meaning the whole thing will typecheck "successfully".

So shouldn't we also check that the mappings returned by `infer_actuals_impl` contain no incoherent binding?

File: contrib/lkt/language/parser.py (hunk 1 above)
Comment (raph-amiard) [reply contrib/lkt/language/parser.py]: Good catch

File: contrib/lkt/language/parser.py (hunk 2 above)
Comment (raph-amiard) [reply contrib/lkt/language/parser.py]: That's correct, there is a missing check, good catch

I'll implement it in a further PR, since globally WIP anyway

**Pull request merged**